plugins/search-backend-node/src/engines/LunrSearchEngine.ts

```diff
@@ -222,3 +239,52 @@ export function decodePageCursor(pageCursor?: string): { page: number } {
 export function encodePageCursor({ page }: { page: number }): string {
   return Buffer.from(`${page}`, 'utf-8').toString('base64');
 }
+
+type ParseHighlightFieldsProps = {
+  preTag: string;
+  postTag: string;
+  doc: any;
+  positionMetadata: {
+    [term: string]: {
+      [field: string]: {
+        position: number[][];
+      };
+    };
+  };
+};
+
+export function parseHighlightFields({
+  preTag,
+  postTag,
+  doc,
+  positionMetadata,
+}: ParseHighlightFieldsProps): { [field: string]: string } {
+  // Merge the field positions across all query terms
+  const highlightFieldPositions = Object.values(positionMetadata).reduce(
+    (fieldPositions, metadata) => {
+      Object.keys(metadata).map(fieldKey => {
+        fieldPositions[fieldKey] = fieldPositions[fieldKey] ?? [];
+        fieldPositions[fieldKey].push(...metadata[fieldKey].position);
+      });
+
+      return fieldPositions;
+    },
+    {} as { [field: string]: number[][] },
+  );
+
+  return Object.fromEntries(
+    Object.entries(highlightFieldPositions).map(([field, positions]) => {
+      positions.sort((a, b) => b[0] - a[0]);
+
+      const highlightedField = positions.reduce((content, pos) => {
+        return (
+          `${content.substring(0, pos[0])}${preTag}` +
+          `${content.substring(pos[0], pos[0] + pos[1])}` +
+          `${postTag}${content.substring(pos[0] + pos[1])}`
```

Comment:
Just double checking:

There might be a need to escape the `content` before wrapping it in the tags. A very simple example would be the documentation about this feature itself, containing an example with named tags.

As I understand it, we currently rely on each engine's implementation to remedy this. So in this case we are using a UUID which sufficiently resolves this issue.

Is there any option to move this one level higher up, i.e. independent of engines?

Just a thought, current implementation looks correct 👍

Reply:
hmm I would not expect the content to ever be rendered directly (ie. require escaping) as HTML. I believe this is also safe-guarded against due to React auto escaping values within JSX and would require explicit calls to `dangerouslySetInnerHTML` to render directly (at which it should be the responsibility of the caller to always ensure the value is properly sanitized)